File: spartan/worker.py
#!/usr/bin/env python

'''
This module defines the `Worker` class and related helper functions.

Workers in Spartan manage array data and computation; methods are 
available for creating, updating, reading and deleting *tiles* of
arrays.  Workers can also run a user-specified function on a set
of tiles.

Workers periodically send a heartbeat message to the master; if the
master cannot be contacted for a sufficiently long interval, workers
shut themselves down.   
'''

import multiprocessing
from multiprocessing.pool import ThreadPool
import os
import pstats
import socket
import sys
import threading
import time

from . import config, util, rpc, core, blob_ctx
from .config import FLAGS, StrFlag, IntFlag, BoolFlag
from .rpc import zeromq, TimeoutException, rlock
from .util import Assert
import psutil
import weakref
import numpy as np
import scipy.sparse
from spartan import sparse

#timeout for hearbeat messsage
HEARTBEAT_TIMEOUT=10
_init_lock = rlock.FastRLock()

class Worker(object):
  '''
  Spartan workers generally correspond to one core of a machine.
  
  Workers manage the storage of array data and running of kernel
  functions.
  
  Attributes:
      id (int): The unique identifier for this worker
      _peers (dict): Mapping from worker id to RPC client
      _blobs (dict): Mapping from tile id to tile.
  '''
  def __init__(self, master):
    # Reseed the Numpy random number state.
    # 
    # The default initialization results in all worker processes on the machine having the 
    # same random seed.
    np.random.seed(seed=os.getpid())

    self.id = -1
    self._initialized = False
    self._peers = {}
    self._blobs = {}
    self._master = master
    self._running = True
    self._ctx = None
    self.worker_status = core.WorkerStatus(psutil.TOTAL_PHYMEM, psutil.NUM_CPUS,
                                           psutil.virtual_memory().percent,
                                           psutil.cpu_percent(), 
                                           time.time(),
                                           [], [])
    
    self._lock = rlock.FastRLock()
    #self._lock = threading.Lock()
    
    #Patch to fix buggy assumption by multiprocessing library  
    if not hasattr(threading.current_thread(), "_children"):
      threading.current_thread()._children = weakref.WeakKeyDictionary()
    
    self._kernel_threads = ThreadPool(processes=1)
    
    if FLAGS.profile_worker:
      import yappi
      yappi.start()

    hostname = socket.gethostname()
    self._server = rpc.listen_on_random_port(hostname)
    self._server.register_object(self)

    if FLAGS.profile_worker:
      self._server._socket._event_loop.enable_profiling()

    self._server.serve_nonblock()

    req = core.RegisterReq()
    req.host = hostname
    req.port = self._server.addr[1]
    req.worker_status = self.worker_status

    with _init_lock:
      # There is a race-condition in the initialization code for zeromq; this causes
      # sporadic crashes when running in multi-thread mode.  We lock the first
      # client RPC to workaround this issue.
      master.register(req)

  def initialize(self, req, handle):
    '''
    Initialize worker.
    
    Assigns this worker a unique identifier and sets up connections to all other workers in the process.
    
    Args:
        req (InitializeReq): foo
        handle (PendingRequest): bar
    
    '''
    util.log_debug('Worker %d initializing...', req.id)
    for id, (host, port) in req.peers.iteritems():
      self._peers[id] = rpc.connect(host, port)
    self._peers[blob_ctx.MASTER_ID] = self._master
    
    self.id = req.id
    self._ctx = blob_ctx.BlobCtx(self.id, self._peers, self)
    blob_ctx.set(self._ctx)
    self._initialized = True
    handle.done()
    
  def create(self, req, handle):
    '''
    Create a new tile.
    
    :param req: `CreateTileReq`
    :param handle: `PendingRequest`
    
    '''
    with self._lock:
      assert self._initialized
      #util.log_info('Creating: %s', req.tile_id)
      Assert.eq(req.tile_id.worker, self.id)
  
      if req.tile_id.id == -1:
        id = self._ctx.new_tile_id()
      else:
        id = req.tile_id
  
      self._blobs[id] = req.data
      resp = core.CreateTileResp(tile_id=id)
    handle.done(resp)

  def tile_op(self, req, handle):
    resp = core.RunKernelResp(result=req.fn(self._blobs[req.tile_id]))
    handle.done(resp)
        
  def destroy(self, req, handle):
    '''
    Delete zero or more blobs.
    
    :param req: `DestroyReq`
    :param handle: `PendingRequest`
    
    '''
    with self._lock:
      for id in req.ids:
        if id in self._blobs:
          del self._blobs[id]
          #util.log_info('Destroyed blob %s', id)

    #util.log_info('Destroy...')
    handle.done()

  def update(self, req, handle):
    '''
    Apply an update to a tile.
    
    :param req: `UpdateReq`
    :param handle: `PendingRequest`
    
    '''
    #util.log_info('W%d Update: %s', self.id, req.id)
    with self._lock:
      blob =  self._blobs[req.id]
      self._blobs[req.id] = blob.update(req.region, req.data, req.reducer)
    
    handle.done()

  def get(self, req, handle):
    '''
    Fetch a portion of a tile.
    
    :param req: `GetReq`
    :param handle: `PendingRequest`
    
    '''
    if req.subslice is None:
      #util.log_info('GET: %s', type(self._blobs[req.id]))
      resp = core.GetResp(data=self._blobs[req.id])
      handle.done(resp)
    else:
      resp = core.GetResp(data=self._blobs[req.id].get(req.subslice))
      handle.done(resp)

  def get_flatten(self, req, handle):
    '''
    Fetch a flatten portion of the flatten format of a tile.
    
    :param req: `GetReq`
    :param handle: `PendingRequest`
    
    '''
    if req.subslice is None:
      #util.log_info('GET: %s', type(self._blobs[req.id]))
      resp = core.GetResp(data=self._blobs[req.id].data.flatten())
      handle.done(resp)
    else:
      resp = core.GetResp(data=self._blobs[req.id].data.flatten()[req.subslice])
      handle.done(resp)

  def _run_kernel(self, req, handle):
    '''
    Run a kernel over the tiles resident on this worker.
    
    (This operation is run in a separate thread from the main worker loop).
    
    :param req: `KernelReq`
    :param handle: `PendingRequest`
    
    '''
    start_time = time.time()
    futures = []
    try:
      blob_ctx.set(self._ctx)
      results = {}

      local_reduction = sparse.create_unordered_map()
      if 'target' in req.kw and 'local_reduction' in req.kw['fn_kw']:
        req.kw['fn_kw']['local_reduction'] = local_reduction

      for tile_id in req.blobs:
        #util.log_info('%s %s', tile_id, tile_id in self._blobs)
        if tile_id in self._blobs:
          #util.log_info('W%d kernel start', self.id)
          blob = self._blobs[tile_id]
          map_result = req.mapper_fn(tile_id, blob, **req.kw)
          results[tile_id] = map_result.result
          
          if map_result.futures is not None:
            futures.append(map_result.futures)

          
      #util.log_info('W%d kernel finish %f', self.id, time.time() - begin)
      # wait for all kernel update operations to finish
      rpc.wait_for_all(futures)
      if 'target' not in req.kw or 'local_reduction' not in req.kw['fn_kw']:
        handle.done(results)
      else:
        target = req.kw['target']
        ex = results.values()[0]
        tt1 = time.time()
        new_rows, new_cols, new_data  = sparse.generate_reduction_data(local_reduction)
        tt2 = time.time()
        v = scipy.sparse.coo_matrix((new_data, (new_rows, new_cols)), shape=target.shape, dtype=target.dtype)
        tt3 = time.time()
        target.update(ex, v)
        tt4 = time.time()
        util.log_warn('sort:%s, create coo:%s, update:%s', tt2 - tt1, tt3 - tt2, tt4 - tt3)
        handle.done({})
  
      #util.log_warn('Waiting for %s futures done %f', len(futures), time.time() - begin)
    except:
      util.log_warn('Exception occurred during kernel call', exc_info=1)
      self.worker_status.add_task_failure(req)
      handle.exception()
    
    finish_time = time.time()
    self.worker_status.add_task_report(req, start_time, finish_time)

  def run_kernel(self, req, handle):
    '''
    Run a kernel on tiles local to this worker.
    
    :param req: `KernelReq`
    :param handle: `PendingRequest`
    
    '''
    #threading.Thread(target=self._run_kernel, args=(req, handle)).start()
    self._kernel_threads.apply_async(self._run_kernel, args=(req, handle))
      
  def shutdown(self, req, handle):
    '''
    Shutdown this worker.
    
    Shutdown is deferred to another thread to ensure the RPC reply
    is sent before the poll loop is killed.
    
    :param req: `EmptyMessage`
    :param handle: `PendingRequest`
    
    '''
    util.log_info('Shutdown worker %d (profile? %d)', self.id, FLAGS.profile_worker)
    if FLAGS.profile_worker:
      try:
        os.system('mkdir -p ./_worker_profiles/')
        import yappi
        yappi.get_func_stats().save('_worker_profiles/%d' % self.id, type='pstat')
      except Exception, ex:
        print 'Failed to write profile.', ex
    handle.done()
    threading.Thread(target=self._shutdown).start()

  def _shutdown(self):
    util.log_debug('Closing server...')
    time.sleep(0.1)
    self._running = False
    self._server.shutdown()
  
  def wait_for_shutdown(self):
    '''
    Wait for the worker to shutdown.
    
    Periodically send heartbeat updates to the master.
    '''
    last_heartbeat = time.time()
    while self._running:
      now = time.time()
      if now - last_heartbeat < FLAGS.heartbeat_interval or not self._initialized:
        time.sleep(0.1)
        continue
      
      self.worker_status.update_status(psutil.virtual_memory().percent, psutil.cpu_percent(), now)
      future = self._ctx.heartbeat(self.worker_status, HEARTBEAT_TIMEOUT)  
      try:
        future.wait()
      except TimeoutException:
        util.log_info("Exit due to heartbeat message timeout.")
        sys.exit(0)

      last_heartbeat = time.time()

    util.log_info('Worker shutdown.  Exiting.')


def _start_worker(master, local_id):
  '''
  Start a worker, register it with the master process, and wait
  until the worker is shutdown.
  
  Runs in a subprocess.
  
  :param master: (host, port)
  :param local_id: index (from 0..#num_workers) of this worker on the local machine.
  '''
  util.log_info('Worker starting up... Master: %s Profile: %s', master, FLAGS.profile_worker)
  rpc.set_default_timeout(FLAGS.default_rpc_timeout)
  if FLAGS.use_single_core:
    pid = os.getpid()
    os.system('taskset -pc %d %d > /dev/null' % (local_id * 2, pid))
    
  master = rpc.connect(*master)
  worker = Worker(master)
  worker.wait_for_shutdown()

  if FLAGS.dump_timers:
    util.TIMER.dump()

if __name__ == '__main__':
  sys.path.append('./tests')

  FLAGS.add(StrFlag('master'))
  FLAGS.add(IntFlag('count'))
  FLAGS.add(IntFlag('heartbeat_interval'))

  #resource.setrlimit(resource.RLIMIT_AS, (8 * 1000 * 1000 * 1000,
  #                                        8 * 1000 * 1000 * 1000))

  config.parse(sys.argv)
  assert FLAGS.master is not None
  assert FLAGS.count > 0
  assert FLAGS.heartbeat_interval > 0

  util.log_info('Starting %d workers on %s', FLAGS.count, socket.gethostname())

  m_host, m_port = FLAGS.master.split(':')
  master = (m_host, int(m_port))

  workers = []
  for i in range(FLAGS.count):
    p = multiprocessing.Process(target=_start_worker, 
                                args=(master, i))
    p.start()
    workers.append(p)
    
    
  def kill_workers():
    for p in workers:
      p.terminate()
      
  watchdog = util.FileWatchdog(on_closed=kill_workers)
  watchdog.start()
  
  for w in workers:
    w.join()
    
  print >>sys.stderr, 'Worker: all worker processes exited!'

File: spartan/worker.cc
#include <string>
#include <stdio.h>
#include <signal.h>
#include <sys/wait.h>

#include "cconfig.h"
#include "cblob_ctx.h"
#include "cworker.h"

class GILHelper {
    PyGILState_STATE gil_state;
public:
    GILHelper() {
        gil_state = PyGILState_Ensure();
    }

    ~GILHelper() {
        PyGILState_Release(gil_state);
    }
};

CWorker::CWorker(const std::string& master_addr, const std::string& worker_addr,
               int32_t heartbeat_interval) {
    id = -1;
    _id_counter = 0;
    _addr = worker_addr;
    _initialized = false;
    _running = true;
    _ctx = NULL;
    _clt_poll = new rpc::PollMgr;
    _clt_pool = new rpc::ClientPool(_clt_poll);
    _master = new spartan::MasterProxy(_clt_pool->get_client(master_addr));
    _worker_status = new WorkerStatus(sysconf(_SC_PHYS_PAGES) * sysconf(_SC_PAGE_SIZE),
                                      (int32_t)sysconf(_SC_NPROCESSORS_CONF),
                                      0, 0, (double)time(0), _kernel_remain_tiles);
    HEARTBEAT_INTERVAL = heartbeat_interval;
}

CWorker::~CWorker() {
    for (auto& it : _peers) {
        delete it.second;
    }
    _peers.clear();
    delete _ctx;

    delete _master;
    delete _worker_status;

    delete _clt_pool;
    _clt_poll->release();
}

void CWorker::register_to_master() {
    RegisterReq req(_addr, *_worker_status);
    rpc::FutureAttr fu_attr;
    fu_attr.callback = [this] (rpc::Future* fu) {
        if (fu->get_error_code() != 0) {
            Log_error("Exit due to register message error:%d.", fu->get_error_code());
            this->_running = false;
        }
        fu->release();
    };
    rpc::Future* fu = _master->async_reg(req, fu_attr);
    if (fu == NULL) {
        Log_error("Exit due to connection to master failed.");
        _running = false;
    }
}

void CWorker::wait_for_shutdown() {
    double last_heartbeat = (double)time(0);
    while (_running) {
        double now = (double)time(0);
        if (now - last_heartbeat < HEARTBEAT_INTERVAL || !_initialized) {
            sleep(0.5);
            continue;
        }

        _worker_status->update_status(0, 0, now);
        HeartbeatReq req(id, *_worker_status);

        rpc::FutureAttr fu_attr;
        fu_attr.callback = [this] (rpc::Future* fu) {
            if (fu->get_error_code() == ETIMEDOUT) {
                Log_error("Exit due to heartbeat message timeout.");
                this->_running = false;
            }
        };
        rpc::Future* fu = _master->async_heartbeat(req, fu_attr);
        if (fu == NULL) {
            Log_error("Exit due to connection to master failed.");
            this->_running = false;
            break;
        }
        fu->timed_wait(1);
        fu->release();

        last_heartbeat = (double)time(0);
    }
    Log_info("CWorker %d shutdown. Exiting.", id);
}

/**
 * Initialize worker.
 * Assigns this worker a unique identifier and sets up connections to all other workers in the process.
 *
 * Args:
 *     req (InitializeReq): foo
 *     resp (EmptyMessage): bar
 */

void CWorker::initialize(const InitializeReq& req, EmptyMessage* resp) {
    Log_debug("CWorker %d initializing...", req.id);
    id = req.id;

    for (auto& it : req.peers) {
        _peers[it.first] = new spartan::WorkerProxy(_clt_pool->get_client(it.second));
    }
    _ctx = new CBlobCtx(id, &_peers, this);
    _initialized = true;
    Log_info("Worker %d initialization done!", id);
}

void CWorker::get_tile_info(const TileIdMessage& req, TileInfoResp* resp) {
    lock(_blob_lock);
    std::unordered_map<TileId, CTile>::iterator it = _blobs.find(req.tile_id);
    unlock(_blob_lock);
    assert(it != _blobs.end());
    resp->dtype = (it->second).dtype;
    resp->sparse = ((it->second).type == CTile::TYPE_SPARSE);
}

void CWorker::create(const CreateTileReq& req, TileIdMessage* resp) {
    resp->tile_id.worker = id;
    resp->tile_id.id = _id_counter++;
    lock(_blob_lock);
    _blobs[resp->tile_id] = req.data;
    unlock(_blob_lock);
}

void CWorker::destroy(const DestroyReq& req, EmptyMessage* resp) {
    lock(_blob_lock);
    for (auto& tid : req.ids) {
        Log_debug("destroy tile:%s", tid.to_string().c_str());
        if (tid.worker == id)
            _blobs.erase(tid);
    }
    unlock(_blob_lock);
}

void CWorker::update(const UpdateReq& req, EmptyMessage* resp) {
    lock(_blob_lock);
    std::unordered_map<TileId, CTile>::iterator it = _blobs.find(req.id);
    unlock(_blob_lock);
    assert(it != _blobs.end());
    // TODO: update tile data
    //it->second.update(req.region, req.data, req.reducer);
}

void CWorker::get(const GetReq& req, GetResp* resp) {
    Log_debug("receive get %s[%d:%d:%d]", req.id.to_string().c_str(),
              req.subslice.slices[0].start, req.subslice.slices[0].stop, req.subslice.slices[0].step);
    resp->data = "get success! in worker " + std::to_string(id);

    lock(_blob_lock);
    std::unordered_map<TileId, CTile>::iterator it = _blobs.find(req.id);
    unlock(_blob_lock);
    assert(it != _blobs.end());
    // TODO: get tile data
    // resp->data = it->second.get(req.subslice);
}

void CWorker::get_flatten(const GetReq& req, GetResp* resp) {
    Log_debug("receive get_flatten %s[%d:%d:%d]", req.id.to_string().c_str(),
              req.subslice.slices[0].start, req.subslice.slices[0].stop, req.subslice.slices[0].step);

    lock(_blob_lock);
    std::unordered_map<TileId, CTile>::iterator it = _blobs.find(req.id);
    unlock(_blob_lock);
    assert(it != _blobs.end());
    // TODO: get tile data
    // resp->data = it->second.get(req.subslice);

}

void CWorker::cancel_tile(const TileIdMessage& req, rpc::i8* resp) {
    Log_info("receive cancel_tile %s", req.tile_id.to_string().c_str());
    lock(_kernel_lock);
    if (_kernel_remain_tiles.size() > 0) {
        assert(_kernel_remain_tiles.front() == req.tile_id);
        _kernel_remain_tiles.erase(_kernel_remain_tiles.begin());
        *resp = 1;
    } else {
        *resp = 0;
    }
    unlock(_kernel_lock);
}

void CWorker::run_kernel(const RunKernelReq& req, RunKernelResp* resp) {
    Log_debug("receive run_kernel req");

    for (auto& tid : req.blobs) {
        if (tid.worker == id)
            _kernel_remain_tiles.push_back(tid);
    }

    //if (_kernel_remain_tiles.size() == 0) return;

    // TODO:sort _kernel_remain_tiles according to tile size
    // self._kernel_remain_tiles.sort()

    char init_cmd[] = "blob_ctx.set(blob_ctx.WorkerBlobCtx(worker_ctx))\n"
                      "mapper_fn, kw = read(fn)\n"
                      "results={}\n"
                      "futures=FutureGroup()\n";

    char mapper_cmd[] = "tile_id = core.TileId(*tid)\n"
        "map_result = mapper_fn(tile_id, blob, **kw)\n"
        "results[tile_id]=map_result.result\n"
        "if map_result.futures is not None:\n\tfutures.append(map_result.futures)\n";

    PyObject *pMain, *pLocal;
    {
        GILHelper gil_helper;

        pMain = PyImport_AddModule("__main__");
        pLocal = PyModule_GetDict(pMain);

        PyRun_SimpleString("from spartan import blob_ctx, core");
        PyRun_SimpleString("from spartan.fastrpc import read, serialize, FutureGroup");

        PyDict_SetItemString(pLocal, "worker_ctx", Py_BuildValue("k", _ctx));
        PyDict_SetItemString(pLocal, "fn", Py_BuildValue("s", req.fn.c_str()));

        PyRun_String(init_cmd, Py_file_input, pLocal, pLocal);
        PyRun_SimpleString("print mapper_fn, kw\nprint mapper_fn(*kw)\n");
        PyRun_SimpleString("ctx = blob_ctx.get()\n");
    }

    while (_kernel_remain_tiles.size() > 0) {
        lock(_kernel_lock);
        TileId& tid = _kernel_remain_tiles.back();
        _kernel_remain_tiles.pop_back();
        unlock(_kernel_lock);

        //CTile blob = _blobs[tid];

        Log_debug("process tile:%s", tid.to_string().c_str());
        {
            GILHelper gil_helper;

            /*
            PyDict_SetItemString(pLocal, "blob", Py_BuildValue("s", blob));
            PyDict_SetItemString(pLocal, "tid", Py_BuildValue("(ii)", tid.worker, tid.id));
            PyRun_String(mapper_cmd, Py_file_input, pLocal, pLocal);
            */
            PyDict_SetItemString(pLocal, "tid", Py_BuildValue("(ii)", tid.worker, tid.id));
            PyRun_SimpleString("results[core.TileId(*tid)] = ctx.get(core.TileId(0,4), [slice(0,1,2)])\n");
        }
    }

    {
        GILHelper gil_helper;
        PyRun_SimpleString("futures.wait()\n");
    }

    // TODO check load balance
    /*
      # We've finished processing our local set of tiles.
      # If we are load balancing, check with the master if it's possible to steal
      # a tile from another worker.
      if FLAGS.load_balance:
        tile_id = self._ctx.maybe_steal_tile(None, None).tile_id
        while tile_id is not None:
          blob = self._ctx.get(tile_id, None)
          map_result = req.mapper_fn(tile_id, blob, **req.kw)
          with self._lock:
            id = self._ctx.new_tile_id()
            self._blobs[id] = blob
          results[id] = map_result.result
          if map_result.futures is not None:
            rpc.wait_for_all(map_result.futures)

          tile_id = self._ctx.maybe_steal_tile(tile_id, id).tile_id
    */
    {
        GILHelper gil_helper;
        PyRun_String("returnstr = serialize(results)\n", Py_file_input, pLocal, pLocal);
        PyObject* re = PyDict_GetItemString(pLocal, "returnstr");
        resp->result = std::string(PyString_AsString(re), PyString_Size(re));
    }
}

void start_worker(int32_t port, int argc, char** argv) {
    std::string w_addr = "0.0.0.0:" + std::to_string(port);
    Log_info("start worker pid %d at %s", getpid(), w_addr.c_str());

    CWorker* w = new CWorker(((StrFlag*)FLAGS.get("master"))->get(), w_addr,
                             ((IntFlag*)FLAGS.get("heartbeat_interval"))->get());

    // start rpc server
    rpc::PollMgr* poll = new rpc::PollMgr;
    rpc::ThreadPool* pool = new rpc::ThreadPool(2);
    rpc::Server *server = new rpc::Server(poll, pool);
    server->reg(w);

    if (server->start(w_addr.c_str()) == 0) {
        // init python environment
        Py_Initialize();
        PyEval_InitThreads();
        PyEval_ReleaseThread(PyThreadState_Get());

        {
            GILHelper gil_helper;
            PySys_SetArgv(argc, argv);
            PyRun_SimpleString("import sys");
            PyRun_SimpleString("import spartan");
            PyRun_SimpleString("spartan.config.parse(sys.argv)");
        }

        w->register_to_master();
        w->wait_for_shutdown();

        // finalize python interpreter
        PyGILState_Ensure();
        Py_Finalize();
    }

    delete server;
    pool->release();
    poll->release();
    delete w;
}

int main(int argc, char* argv[]) {
    config_parse(argc, argv);

    base::LOG_LEVEL = ((LogLevelFlag*)FLAGS.get("log_level"))->get();
    int num_workers = ((IntFlag*)FLAGS.get("count"))->get();
    int port_base = ((IntFlag*)FLAGS.get("port_base"))->get() + 1;

    // start #num_workers workers in this host
    for (int i = 0; i < num_workers; i++) {
        if (fork() == 0) {
            start_worker(port_base + i, argc, argv);
            return 0;
        }
    }

    pid_t pid;
    while ((pid = wait(NULL)) >= 0) {
        Log_info("child pid %d finish!", pid);
    }

    return 0;
}